File: Serialize_deserialize.java

//TC - O(N)
//SC - O(N)
import java.util.*;

class Codec {

	// Encodes a tree to a single string.
	public String serialize(TreeNode root) {
		if (root == null)
			return "";
		StringBuilder sb = new StringBuilder();
		Queue<TreeNode> q = new LinkedList<>();
		q.add(root);
		while (!q.isEmpty()) {
			TreeNode curr = q.poll();
			if (curr != null) {
				sb.append(curr.val);
				sb.append(",");
				q.add(curr.left);
				q.add(curr.right);
			} else {
				sb.append("null");
				sb.append(",");
			}
		}
		return sb.toString();
	}

	// Decodes your encoded data to tree.
	public TreeNode deserialize(String data) {
		if (data == "")
			return null;
		String[] strArr = data.split(",");
		Queue<TreeNode> q = new LinkedList<>();
		int i = 0;
		TreeNode root = new TreeNode(Integer.parseInt(strArr[i]));
		q.add(root);
		i++;
		while (!q.isEmpty() && i < strArr.length) {
			TreeNode curr = q.poll();
			if (!strArr[i].equals("null")) {
				curr.left = new TreeNode(Integer.parseInt(strArr[i]));
				q.add(curr.left);
			}
			i++;
			if (!strArr[i].equals("null")) {
				curr.right = new TreeNode(Integer.parseInt(strArr[i]));
				q.add(curr.right);
			}
			i++;
		}
		return root;
	}
}

File: serialize_deserialize.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        q = []
        q.append(root)
        r = ""
        while q:
            curr = q.pop(0)
            if curr == "Null":
                r += "Null"
                r += ","
            else:
                r += str(curr.val)
                r += ","
                if curr.left:
                    q.append(curr.left)
                else:
                    q.append("Null")
                if curr.right:
                    q.append(curr.right)
                else:
                    q.append("Null")
        return r

    def deserialize(self, data):
        if not data  or len(data) == 0 :
            return None 
        s = data.split(",")
        root_val = int(s[0])
        idx = 1
        q = []
        root = TreeNode(root_val)
        q.append(root)
        while q and idx < len(s):
            curr = q.pop(0)
            if not s[idx] == "Null":
                curr.left = TreeNode(s[idx])
                q.append(curr.left)
            idx += 1

            if not s[idx] == "Null":
                curr.right = TreeNode(s[idx])
                q.append( curr.right )
            idx += 1
        return root
        
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
